go/libraries/doltcore/sqle/stats/stats_provider.go

```diff
@@ -164,32 +172,83 @@ type indexMeta struct {
 	cols  []string
 }
 
-type statsMeta struct {
-	db    string
-	table string
-	index string
-}
-
 func NewProvider() *Provider {
 	return &Provider{
-		//indexToStats: make(map[indexMeta][]statsMeta),
-		stats: make(map[sql.StatQualifier]*DoltStats),
+		mu:      &sync.Mutex{},
+		dbStats: make(map[string]*dbStats),
 	}
 }
 
+// Provider is the engine interface for reading and writing index statistics.
+// Each database has its own statistics table that all tables/indexes in a db
+// share.
 type Provider struct {
+	mu             *sync.Mutex
 	latestRootAddr hash.Hash
-	//indexToStats   map[indexMeta][]statsMeta
-	stats map[sql.StatQualifier]*DoltStats
+	dbStats        map[string]*dbStats
+}
+
+// each database has one statistics table that is a collection of the
+// table stats in the database
+type dbStats struct {
+	db     string
+	active map[hash.Hash]int
+	stats  map[sql.StatQualifier]*DoltStats
 }
 
 var _ sql.StatsProvider = (*Provider)(nil)
 
+// Init scans the statistics tables, populating the |stats| attribute.
+// Statistics are not available for reading until we've finished loading.
+func (p *Provider) Load(ctx *sql.Context, dbNames []string, dbs []*doltdb.DoltDB) error {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	eg, ctx := ctx.NewErrgroup()
+	for i, db := range dbs {
+		// copy closure variables
+		dbName := dbNames[i]
+		closureDb := db
+		eg.Go(func() (err error) {
+			defer func() {
+				if r := recover(); r != nil {
+					if str, ok := r.(fmt.Stringer); ok {
+						err = fmt.Errorf("%w: %s", ErrFailedToLoad, str.String())
+					} else {
+						err = fmt.Errorf("%w: %v", ErrFailedToLoad, r)
+					}
+
+					return
+				}
+			}()
+
+			m, err := closureDb.GetStatistics(ctx)
+			if errors.Is(err, doltdb.ErrNoStatistics) {
+				return nil
+			} else if err != nil {
+				return err
+			}
+			stats, err := loadStats(ctx, dbName, m)
+			if errors.Is(err, ErrIncompatibleVersion) {
+				ctx.Warn(0, err.Error())
+				return nil
+			} else if err != nil {
+				return err
+			}
+			p.dbStats[dbName] = stats
```

Comment:
It's not safe to modify maps concurrently: https://go.dev/blog/maps#TOC_6.

The official suggestion here is to use a mutex, but I suspect it would also be safe to insert nil values for each db name before creating any of the goroutines, and then reassigning those values in the goroutines. That should guarentee that the map doesn't get resized while another routine is writing to it.

Reply:
I think I misread; if p.dbStats[dbName] already exists for each dbName this is probably safe.